File: scala/egcd/src/main/scala/de/csmath/math/Egcd.scala
package de.csmath.math;

object Egcd {

    //
    // egcd(m,n) = (s,t,d) such that
    // d = gcd(m,n) and
    // sm + tn = d
    //
    def egcd(m: Int, n: Int) = egcdaux(1, 0, 0, 1, m, n)

    //
    // let a', b', a, b, m, n, c, d be integers such that
    //
    //    a'm + b'n = c   and
    //    am  + bn  = d
    //
    //              a' b' a  b  c  d
    // then egcdaux(1, 0, 0, 1, m, n) = (s,t,d) such that
    // d = gcd(m,n) and
    // sm + tn = d
    //
    private def egcdaux(a1: Int, b1: Int, a: Int, b: Int, c: Int, d: Int): (Int,Int,Int) = {
        val q = c / d
        val r = c % d
        if (r==0)
           (a,b,d)
        else egcdaux(a, b, a1 - q*a, b1 - q*b, d, r)
    }

}

File: javascript/egcd.js

function egcd(m,n) {
    var a1 = 1;
    var b1 = 0;
    var a  = 0;
    var b  = 1;
    var c  = m;
    var d  = n;
    var q = Math.floor(c/d);
    var r = c % d;
    // such that
    //             a1 m + b2 n = c
    //                 am + bn = d
    // (a1 - qa)m + (b1 - qb)n = r
    while (r > 0) {
        var t = a1;
        a1 = a;
        a = t - q*a; // a1 - qa
        t = b1;
        b1 = b;
        b = t - q*b; // b1 - qb
        c = d;
        d = r;
        q = Math.floor(c/d);
        r = c % d;
    }
    return [a,b,d];
}
